File: FRONTEND_NEXTJS/src/lib/document-processing.ts
import JSZip from "jszip";

export type ExtractedPage = {
  pageNumber?: number;
  text: string;
};

export async function extractTextFromBuffer(params: {
  buffer: Buffer;
  mimeType: string;
  fileName: string;
}): Promise<ExtractedPage[]> {
  const fileName = params.fileName.toLowerCase();

  if (params.mimeType.includes("pdf") || fileName.endsWith(".pdf")) {
    const pdfParse = (await import("pdf-parse")).default;
    const parsed = await pdfParse(params.buffer);
    return [{ text: normalizeText(parsed.text) }];
  }

  if (
    params.mimeType.includes("wordprocessingml") ||
    fileName.endsWith(".docx")
  ) {
    const mammoth = await import("mammoth");
    const result = await mammoth.extractRawText({ buffer: params.buffer });
    return [{ text: normalizeText(result.value) }];
  }

  if (
    params.mimeType.includes("presentationml") ||
    fileName.endsWith(".pptx")
  ) {
    return extractPptxText(params.buffer);
  }

  if (params.mimeType.startsWith("text/") || fileName.endsWith(".txt")) {
    return [{ text: normalizeText(params.buffer.toString("utf8")) }];
  }

  throw new Error("Unsupported file type. Upload PDF, PPTX, DOCX, or TXT.");
}

export function chunkExtractedPages(pages: ExtractedPage[]) {
  const chunks: Array<{ content: string; pageNumber?: number; chunkIndex: number }> = [];
  let chunkIndex = 0;

  for (const page of pages) {
    const words = page.text.split(/\s+/).filter(Boolean);
    const size = 650;
    const overlap = 90;

    for (let start = 0; start < words.length; start += size - overlap) {
      const content = sanitizeTextForStorage(words.slice(start, start + size).join(" "));
      if (content.length < 40) continue;

      chunks.push({
        content,
        pageNumber: page.pageNumber,
        chunkIndex
      });
      chunkIndex += 1;
    }
  }

  return chunks;
}

async function extractPptxText(buffer: Buffer): Promise<ExtractedPage[]> {
  const zip = await JSZip.loadAsync(buffer);
  const slideFiles = Object.keys(zip.files)
    .filter((name) => /^ppt\/slides\/slide\d+\.xml$/.test(name))
    .sort((a, b) => getSlideNumber(a) - getSlideNumber(b));

  const pages: ExtractedPage[] = [];

  for (const slidePath of slideFiles) {
    const xml = await zip.files[slidePath].async("string");
    const text = Array.from(xml.matchAll(/<a:t>(.*?)<\/a:t>/g))
      .map((match) => decodeXml(match[1]))
      .join(" ");

    pages.push({
      pageNumber: getSlideNumber(slidePath),
      text: normalizeText(text)
    });
  }

  return pages;
}

function getSlideNumber(path: string) {
  return Number(path.match(/slide(\d+)\.xml$/)?.[1] || 0);
}

export function sanitizeTextForStorage(text: string) {
  return text
    .replace(/\u0000/g, "")
    .replace(/[\u0001-\u0008\u000B\u000C\u000E-\u001F\u007F]/g, " ")
    .replace(/\uFFFD/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

function normalizeText(text: string) {
  return sanitizeTextForStorage(text);
}

function decodeXml(value: string) {
  return value
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, "\"")
    .replace(/&apos;/g, "'");
}

File: fastapi_backend/services/document_processing.py
import io
import re
from typing import List, Optional
from dataclasses import dataclass


@dataclass
class ExtractedPage:
    text: str
    page_number: Optional[int] = None


def _normalize_text(text: str) -> str:
    return _sanitize_text(text)


def _sanitize_text(text: str) -> str:
    # Remove null bytes and control characters
    text = re.sub(r'\x00', '', text)
    text = re.sub(r'[\x01-\x08\x0b\x0c\x0e-\x1f\x7f]', ' ', text)
    text = re.sub(r'\ufffd', ' ', text)
    text = re.sub(r'\s+', ' ', text)
    return text.strip()


def sanitize_text_for_storage(text: str) -> str:
    return _sanitize_text(text)


def extract_text_from_buffer(
    buffer: bytes,
    mime_type: str,
    file_name: str,
) -> List[ExtractedPage]:
    fname = file_name.lower()

    if "pdf" in mime_type or fname.endswith(".pdf"):
        return _extract_pdf(buffer)

    if "wordprocessingml" in mime_type or fname.endswith(".docx"):
        return _extract_docx(buffer)

    if "presentationml" in mime_type or fname.endswith(".pptx"):
        return _extract_pptx(buffer)

    if mime_type.startswith("text/") or fname.endswith(".txt"):
        return [ExtractedPage(text=_normalize_text(buffer.decode("utf-8", errors="replace")))]

    raise ValueError("Unsupported file type. Upload PDF, PPTX, DOCX, or TXT.")


def _extract_pdf(buffer: bytes) -> List[ExtractedPage]:
    import PyPDF2
    reader = PyPDF2.PdfReader(io.BytesIO(buffer))
    pages = []
    for i, page in enumerate(reader.pages, start=1):
        text = page.extract_text() or ""
        pages.append(ExtractedPage(text=_normalize_text(text), page_number=i))
    return pages


def _extract_docx(buffer: bytes) -> List[ExtractedPage]:
    import docx
    doc = docx.Document(io.BytesIO(buffer))
    text = "\n".join(p.text for p in doc.paragraphs)
    return [ExtractedPage(text=_normalize_text(text))]


def _extract_pptx(buffer: bytes) -> List[ExtractedPage]:
    from pptx import Presentation
    prs = Presentation(io.BytesIO(buffer))
    pages = []
    for i, slide in enumerate(prs.slides, start=1):
        texts = []
        for shape in slide.shapes:
            if shape.has_text_frame:
                for para in shape.text_frame.paragraphs:
                    line = " ".join(run.text for run in para.runs)
                    if line.strip():
                        texts.append(line)
        pages.append(ExtractedPage(
            text=_normalize_text(" ".join(texts)),
            page_number=i,
        ))
    return pages


def chunk_extracted_pages(pages: List[ExtractedPage]):
    """Split pages into overlapping word-level chunks."""
    chunks = []
    chunk_index = 0
    size = 650
    overlap = 90

    for page in pages:
        words = [w for w in page.text.split() if w]
        start = 0
        while start < len(words):
            content = sanitize_text_for_storage(" ".join(words[start: start + size]))
            if len(content) >= 40:
                chunks.append({
                    "content": content,
                    "page_number": page.page_number,
                    "chunk_index": chunk_index,
                })
                chunk_index += 1
            start += size - overlap

    return chunks
